# HR Analysis

## 1. Data Overview and Data Cleaning

### Data Overview

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
from scipy import stats as st
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import ipywidgets as widgets
from IPython.display import display, clear_output






In [2]:
dataset = pd.read_csv('employee_attrition_dataset_10000.csv')
print(f'Rows: {dataset.shape[0]}')
print(f'Columns: {dataset.shape[1]}')
print()
dataset.head(5)


Rows: 10000
Columns: 26



,Employee_ID,Age,Gender,Marital_Status,Department,Job_Role,Job_Level,Monthly_Income,Hourly_Rate,Years_at_Company,...,Overtime,Project_Count,Average_Hours_Worked_Per_Week,Absenteeism,Work_Environment_Satisfaction,Relationship_with_Manager,Job_Involvement,Distance_From_Home,Number_of_Companies_Worked,Attrition
0,1,58,Male,Single,Finance,Manager,5,7332,81,24,...,No,9,48,16,4,1,1,49,3,No
1,2,48,Female,Divorced,HR,Assistant,4,6069,55,18,...,Yes,9,57,10,4,1,1,25,1,No
2,3,34,Female,Married,Marketing,Manager,4,11485,65,6,...,Yes,3,55,1,1,4,3,21,1,Yes
3,4,27,Female,Divorced,HR,Manager,4,18707,28,12,...,No,9,53,2,3,4,1,46,2,No
4,5,40,Male,Married,HR,Analyst,1,16398,92,3,...,No,1,54,11,1,1,1,43,4,No


In [3]:
print('\n-------Data types---------')
dataset.dtypes


-------Data types---------


Employee_ID                      int64
Age                              int64
Gender                             str
Marital_Status                     str
Department                         str
Job_Role                           str
Job_Level                        int64
Monthly_Income                   int64
Hourly_Rate                      int64
Years_at_Company                 int64
Years_in_Current_Role            int64
Years_Since_Last_Promotion       int64
Work_Life_Balance                int64
Job_Satisfaction                 int64
Performance_Rating               int64
Training_Hours_Last_Year         int64
Overtime                           str
Project_Count                    int64
Average_Hours_Worked_Per_Week    int64
Absenteeism                      int64
Work_Environment_Satisfaction    int64
Relationship_with_Manager        int64
Job_Involvement                  int64
Distance_From_Home               int64
Number_of_Companies_Worked       int64
Attrition                

In [4]:
print('\n-------describe(numeric)---------')
dataset.describe()


-------describe(numeric)---------


,Employee_ID,Age,Job_Level,Monthly_Income,Hourly_Rate,Years_at_Company,Years_in_Current_Role,Years_Since_Last_Promotion,Work_Life_Balance,Job_Satisfaction,Performance_Rating,Training_Hours_Last_Year,Project_Count,Average_Hours_Worked_Per_Week,Absenteeism,Work_Environment_Satisfaction,Relationship_with_Manager,Job_Involvement,Distance_From_Home,Number_of_Companies_Worked
count,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000
mean,5000.50000,39.561800,2.990800,11436.716700,57.032300,14.936200,7.451300,4.471900,2.502400,3.038000,2.507700,49.588900,4.984400,44.473500,9.410200,2.493100,2.491400,2.505400,25.272000,2.516600
std,2886.89568,11.454986,1.410643,4926.528302,24.703261,8.431657,4.042903,2.891617,1.112348,1.414764,1.120833,28.801393,2.580043,8.611662,5.760335,1.120927,1.115101,1.116018,14.219474,1.113934
min,1.00000,20.000000,1.000000,3000.000000,15.000000,1.000000,1.000000,0.000000,1.000000,1.000000,1.000000,0.000000,1.000000,30.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000
25%,2500.75000,30.000000,2.000000,7182.500000,36.000000,8.000000,4.000000,2.000000,2.000000,2.000000,2.000000,25.000000,3.000000,37.000000,4.000000,1.000000,1.000000,2.000000,13.000000,2.000000
50%,5000.50000,40.000000,3.000000,11401.500000,57.000000,15.000000,7.000000,4.000000,2.000000,3.000000,3.000000,49.000000,5.000000,45.000000,9.000000,2.000000,2.000000,3.000000,25.000000,2.000000
75%,7500.25000,49.000000,4.000000,15679.500000,78.000000,22.000000,11.000000,7.000000,3.000000,4.000000,4.000000,75.000000,7.000000,52.000000,14.000000,4.000000,3.000000,3.000000,37.000000,4.000000
max,10000.00000,59.000000,5.000000,19999.000000,99.000000,29.000000,14.000000,9.000000,4.000000,5.000000,4.000000,99.000000,9.000000,59.000000,19.000000,4.000000,4.000000,4.000000,49.000000,4.000000


In [5]:
dataset.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   Employee_ID                    10000 non-null  int64
 1   Age                            10000 non-null  int64
 2   Gender                         10000 non-null  str  
 3   Marital_Status                 10000 non-null  str  
 4   Department                     10000 non-null  str  
 5   Job_Role                       10000 non-null  str  
 6   Job_Level                      10000 non-null  int64
 7   Monthly_Income                 10000 non-null  int64
 8   Hourly_Rate                    10000 non-null  int64
 9   Years_at_Company               10000 non-null  int64
 10  Years_in_Current_Role          10000 non-null  int64
 11  Years_Since_Last_Promotion     10000 non-null  int64
 12  Work_Life_Balance              10000 non-null  int64
 13  Job_Satisfaction            

### Data Cleaning

In [6]:
df = dataset.copy()

print('Missing value per column')
print(df.isnull().sum()[df.isnull().sum() > 0] 
      if df.isnull().sum().sum() > 0
       else 'No missing values' )

Missing value per column
No missing values


In [7]:
df.columns = df.columns.str.lower().str.replace(' ', '_')

In [8]:
df.columns

Index(['employee_id', 'age', 'gender', 'marital_status', 'department',
       'job_role', 'job_level', 'monthly_income', 'hourly_rate',
       'years_at_company', 'years_in_current_role',
       'years_since_last_promotion', 'work_life_balance', 'job_satisfaction',
       'performance_rating', 'training_hours_last_year', 'overtime',
       'project_count', 'average_hours_worked_per_week', 'absenteeism',
       'work_environment_satisfaction', 'relationship_with_manager',
       'job_involvement', 'distance_from_home', 'number_of_companies_worked',
       'attrition'],
      dtype='str')

In [9]:
print(f"Duplicated rows across whole dataset: {df.duplicated().sum()}")

print(f'Duplicated in employee_id colimn : {df['employee_id'].duplicated().sum()}')

Duplicated rows across whole dataset: 0
Duplicated in employee_id colimn : 0


In [10]:
text_cols = ['gender', 'marital_status', 'department', 'job_role', 'overtime', 'attrition']

for col in text_cols:
    df[col] = df[col].str.lower()

df['attrition_flag'] = df['attrition'].apply(lambda x: 1 if x == 'yes' else 0)

print(df.shape)
df.sample(10)

(10000, 27)


,employee_id,age,gender,marital_status,department,job_role,job_level,monthly_income,hourly_rate,years_at_company,...,project_count,average_hours_worked_per_week,absenteeism,work_environment_satisfaction,relationship_with_manager,job_involvement,distance_from_home,number_of_companies_worked,attrition,attrition_flag
7448,7449,40,female,single,finance,analyst,4,4340,22,17,...,9,44,13,2,1,4,29,3,yes,1
4572,4573,46,female,married,hr,analyst,5,10304,28,8,...,3,46,10,2,4,3,37,2,no,0
9048,9049,33,female,single,sales,assistant,3,3596,22,1,...,7,51,15,3,4,2,30,2,no,0
2709,2710,30,female,divorced,sales,assistant,3,11551,45,6,...,6,58,10,4,3,2,4,1,no,0
4702,4703,56,male,single,sales,assistant,3,3637,50,28,...,8,59,1,1,1,1,24,3,no,0
2242,2243,30,male,married,finance,analyst,5,8568,99,27,...,4,35,7,3,2,4,24,1,yes,1
5620,5621,44,female,divorced,hr,executive,3,8422,75,4,...,4,55,3,4,4,4,49,1,no,0
5899,5900,26,male,single,it,manager,4,10097,84,17,...,4,45,1,4,4,4,14,2,no,0
4463,4464,51,male,married,it,analyst,2,18015,49,5,...,9,34,7,2,1,4,25,3,yes,1
4395,4396,58,male,divorced,it,assistant,4,12901,75,22,...,8,58,10,2,1,4,15,2,yes,1


## 2. Attrition Analysis

### Total attrition

In [11]:
attrition_count = df['attrition'].value_counts()
print(attrition_count)

rate = df['attrition_flag'].mean()*100
print(f'Attriton rate: {rate}')


attrition
no     8003
yes    1997
Name: count, dtype: int64
Attriton rate: 19.97


In [12]:
fig_attriotion = px.histogram(df, x='attrition',
        title='Attrition count',
        color='attrition',
        color_discrete_sequence=['#636EFA', '#EF553B'])

fig_attriotion.update_yaxes(title_text='Number of employees')
fig_attriotion.show()

### Attrition by department

In [13]:
dep_rate = (df.groupby('department')['attrition_flag'].mean() * 100).round(2)

print(dep_rate)

department
finance      20.85
hr           19.51
it           20.35
marketing    19.36
sales        19.82
Name: attrition_flag, dtype: float64


In [14]:
fig_dep_rate = px.bar(dep_rate, 
                      title='Attrition Rate by Department',
                      labels={'value': 'Attrition Rate (%)', 'index': 'Department'},
                      color=dep_rate.values,
                      color_continuous_scale='RdYlGn_r')

fig_dep_rate.show()

print(f'Highest attrition department: {dep_rate.idxmax()}')

Highest attrition department: finance


### Attrition by jobe role

In [15]:
job_rate = (df.groupby('job_role')['attrition_flag'].mean() * 100).round(2)

job_rate

job_role
analyst      20.22
assistant    21.43
executive    20.07
manager      18.06
Name: attrition_flag, dtype: float64

In [16]:
fig_job_rate = px.bar(job_rate, 
                      title='Attrition Rate by Job Role',
                      labels={'value': 'Attrition Rate (%)', 'index': 'Job Role'},
                      color=job_rate.values,
                      color_continuous_scale='blues')

fig_job_rate.show()

### Overtime and Attrition

In [17]:
overtime_rate = (df.groupby('overtime')['attrition_flag'].mean() * 100).round(2)

overtime_rate

overtime
no     19.87
yes    20.07
Name: attrition_flag, dtype: float64

In [18]:
fig_overtime = px.pie(overtime_rate, 
                      labels=overtime_rate.index,
                      values=overtime_rate.values,
                      title='Attrition Rate by Overtime',
                      color_discrete_sequence=['#636EFA', '#EF553B'])

fig_overtime.show()

## 3. Employee Groups Analysis

### Gender and Marital Status

In [19]:
gender_rate = (df.groupby('gender')['attrition_flag'].mean() * 100).round(2)

gender_rate


fig_gender_rate = px.bar(gender_rate, 
                         title='Attrition Rate by Gender',
                         labels={'value': 'Attrition Rate (%)', 'index': 'Gender'},
                         color=gender_rate.values,
                         color_continuous_scale=['red','blue'])

fig_gender_rate.show()

In [20]:
marital_rate = (df.groupby('marital_status')['attrition_flag'].mean()*100).round(2)

marital_rate


fig_marital_rate = px.bar(marital_rate, 
                          title='Attrition Rate by Marital Status',
                          labels={'value': 'Attrition Rate (%)', 'index': 'Marital Status'},
                          color=marital_rate.values,
                          color_continuous_scale='viridis')

fig_marital_rate.show()

In [21]:
marital_gender = (
    df.groupby(['marital_status', 'gender'])
      .size()
      .unstack(fill_value=0)
      .reindex(['single', 'married', 'divorced'], fill_value=0)
      .rename_axis(index='marital_status', columns='gender')
      .reset_index()
)

marital_gender = marital_gender[['marital_status', 'female', 'male']]

marital_gender




gender,marital_status,female,male
0,single,1678,1617
1,married,1684,1691
2,divorced,1680,1650


In [22]:
marital_gender_long = marital_gender.melt(
    id_vars='marital_status',
    value_vars=['female', 'male'],
    var_name='gender',
    value_name='employees'
)
fig_marital_gender = px.bar(
    marital_gender_long,
    x='marital_status',
    y='employees',
    color='gender',
    barmode='group',
    category_orders={
        'marital_status': ['single', 'married', 'divorced'],
        'gender': ['female', 'male']
    },
    color_discrete_map={
        'female': '#636EFA',
        'male': '#EF553B'
    },
    title='Employees by Marital Status and Gender',
    labels={
        'marital_status': 'Marital Status',
        'employees': 'Number of Employees',
        'gender': 'Gender'
    }
)

fig_marital_gender.update_layout(
    barmode='stack',
    hovermode='x unified'
)

fig_marital_gender.show()

### Age Groups

In [23]:
bins = [17,25,35,45,55,65]
labels = ['17-25','26-35','35-45','46-55','56-65']

df['age_group'] = pd.cut(df['age'],bins=bins,labels=labels)

In [24]:
age_rate = (df.groupby('age_group')['attrition_flag'].mean() * 100).round(2)

age_rate

age_group
17-25    21.78
26-35    18.23
35-45    19.57
46-55    20.54
56-65    21.26
Name: attrition_flag, dtype: float64

In [25]:
age_rate_df = age_rate.reset_index().rename(
    columns={'index': 'age_group', 'attrition_flag': 'attrition_rate'}
)

fig_age_group = px.bar(
    age_rate_df,
    x='age_group',
    y='attrition_rate',
    title='Attrition Rate by Age Group',
    labels={'age_group': 'Age Group', 'attrition_rate': 'Attrition Rate (%)'},
    color='attrition_rate',
    color_continuous_scale='RdYlGn_r',
    text_auto='.2f'
)

fig_age_group.update_traces(
    marker_line_color='black',
    marker_line_width=1.2,
    textfont_size=12
)

fig_age_group.update_layout(
    template='plotly_white',
    xaxis_title='Age Group',
    yaxis_title='Attrition Rate (%)',
    bargap=0.2
)

fig_age_group.show()

### Monthly Income

In [26]:
df.columns

Index(['employee_id', 'age', 'gender', 'marital_status', 'department',
       'job_role', 'job_level', 'monthly_income', 'hourly_rate',
       'years_at_company', 'years_in_current_role',
       'years_since_last_promotion', 'work_life_balance', 'job_satisfaction',
       'performance_rating', 'training_hours_last_year', 'overtime',
       'project_count', 'average_hours_worked_per_week', 'absenteeism',
       'work_environment_satisfaction', 'relationship_with_manager',
       'job_involvement', 'distance_from_home', 'number_of_companies_worked',
       'attrition', 'attrition_flag', 'age_group'],
      dtype='str')

In [27]:
income = df.groupby('attrition')['monthly_income'].mean()

income

attrition
no     11436.385605
yes    11438.043565
Name: monthly_income, dtype: float64

In [28]:
fig_income = px.box(
    df,
    x='attrition',
    y='monthly_income',
    color='attrition',
    category_orders={'attrition': ['no', 'yes']},
    color_discrete_map={'no': "#A9571C", 'yes': "#970B7D"},
    points=False,
    title='Monthly Income Distribution by Attrition',
    labels={
        'attrition': 'Attrition',
        'monthly_income': 'Monthly Income'
    }
)

fig_income.update_layout(
    template='plotly_white',
    showlegend=False,
    yaxis_tickformat=','
)

fig_income.show()

In [29]:
dep_income = df.groupby(['department','attrition'])['monthly_income'].mean().unstack()

dep_income

attrition,no,yes
department,,
finance,11538.801270,11659.310843
hr,11619.408397,11389.133858
it,11381.256881,11511.471795
marketing,11282.781395,11346.346247
sales,11373.844720,11277.346734


In [30]:
fig_avg_income = px.bar(
    dep_income.reset_index(),
    x='department',
    y=['no', 'yes'],
    barmode='group',
    title='Average Monthly Income by Department and Attrition',
    labels={'value': 'Monthly Income', 'department': 'Department'},
    color_discrete_map={'no': '#636EFA', 'yes': '#EF553B'}
)

fig_avg_income.update_layout(
    template='plotly_white',
    yaxis_tickformat=',',
    hovermode='x unified'
)

fig_avg_income.show()

### Satisfaction metrics

In [31]:
df.columns

Index(['employee_id', 'age', 'gender', 'marital_status', 'department',
       'job_role', 'job_level', 'monthly_income', 'hourly_rate',
       'years_at_company', 'years_in_current_role',
       'years_since_last_promotion', 'work_life_balance', 'job_satisfaction',
       'performance_rating', 'training_hours_last_year', 'overtime',
       'project_count', 'average_hours_worked_per_week', 'absenteeism',
       'work_environment_satisfaction', 'relationship_with_manager',
       'job_involvement', 'distance_from_home', 'number_of_companies_worked',
       'attrition', 'attrition_flag', 'age_group'],
      dtype='str')

In [32]:
mean1 = df.groupby('attrition')['job_satisfaction'].mean()
mean2 = df.groupby('attrition')['work_life_balance'].mean()
mean3 = df.groupby('attrition')['work_environment_satisfaction'].mean()

In [33]:
satisfaction = pd.concat(
    [
        mean1.rename('job_satisfaction'),
        mean2.rename('work_life_balance'),
        mean3.rename('work_environment_satisfaction')
    ],
    axis=1
)

satisfaction

,job_satisfaction,work_life_balance,work_environment_satisfaction
attrition,,,
no,3.036486,2.493940,2.499313
yes,3.044066,2.536304,2.468202


In [34]:
fig_satisfaction = px.bar(
    satisfaction.reset_index(),
    x='attrition',
    y=['job_satisfaction', 'work_life_balance', 'work_environment_satisfaction'],
    barmode='group',
    title='Satisfaction Metrics by Attrition',
    labels={'value': 'Average Score', 'attrition': 'Attrition'},
    color_discrete_map={
        'job_satisfaction': '#636EFA',
        'work_life_balance': '#EF553B',
        'work_environment_satisfaction': '#00CC96'
    }
)

fig_satisfaction.update_layout(
    template='plotly_white',
    hovermode='x unified'
)

fig_satisfaction.show()

### Workload metrics

In [35]:
load1 = df.groupby('attrition')['project_count'].mean()
load2 = df.groupby('attrition')['average_hours_worked_per_week'].mean()
load3 = df.groupby('attrition')['absenteeism'].mean()


In [36]:
workload = pd.concat(
    [load1.rename('project_count'),
     load2.rename('average_hours_worked_per_week'),
     load3.rename('absenteeism')], axis=1
)

workload

,project_count,average_hours_worked_per_week,absenteeism
attrition,,,
no,4.975634,44.484068,9.408722
yes,5.019529,44.431147,9.416124


In [37]:
workload_df = workload.reset_index().rename(columns={'index': 'attrition'})

workload_long = workload_df.melt(
    id_vars='attrition',
    var_name='metric',
    value_name='value'
)

workload_long['metric'] = workload_long['metric'].replace({
    'project_count': 'Project Count',
    'average_hours_worked_per_week': 'Avg Hours/Week',
    'absenteeism': 'Absenteeism'
})

fig_workload = px.bar(
    workload_long,
    x='attrition',
    y='value',
    color='attrition',
    facet_col='metric',
    barmode='group',
    category_orders={
        'attrition': ['no', 'yes'],
        'metric': ['Project Count', 'Avg Hours/Week', 'Absenteeism']
    },
    color_discrete_map={'no': '#636EFA', 'yes': '#EF553B'},
    title='Workload Metrics by Attrition Group',
    labels={'attrition': 'Attrition', 'value': 'Average Value', 'metric': 'Metric'}
)

fig_workload.update_layout(
    template='plotly_white',
    legend_title_text='Attrition',
    height=420,
    width=1100,
    hovermode='x unified'
)

fig_workload.show()

### Experience and Tenure

In [38]:
tenure_cols = ['years_at_company','years_in_current_role','years_since_last_promotion']

for col in tenure_cols:
    means = df.groupby('attrition')[col].mean()
    print(col)
    print(means)
    print('-------------------')

    fig_tenure = px.box(
        df,
        x='attrition',
        y=col,
        color='attrition',
        category_orders={'attrition': ['no', 'yes']},
        color_discrete_map={'no': '#636EFA', 'yes': '#EF553B'},
        title=col,
        labels={'attrition': 'Attrition', 'value': 'Years'}
    )

    fig_tenure.update_layout(template='plotly_white', hovermode='x unified')
    fig_tenure.show()

years_at_company
attrition
no     14.924903
yes    14.981472
Name: years_at_company, dtype: float64
-------------------


years_in_current_role
attrition
no     7.440960
yes    7.492739
Name: years_in_current_role, dtype: float64
-------------------


years_since_last_promotion
attrition
no     4.474697
yes    4.460691
Name: years_since_last_promotion, dtype: float64
-------------------


## 4. Relationships and Final decisions

### Chi-Square Test

In [39]:
table = pd.crosstab(df['department'],df['attrition'])
print(table)

chi2,p,dof,excepted = st.chi2_contingency(table)

print(f'Chi2: {chi2}')
print(f'P-value: {p}')

if p < 0.05:
    print("Department and Attrition are related (p < 0.05).")
else:
    print('There is no strong relation between Department and Atrrition')

attrition     no  yes
department           
finance     1575  415
hr          1572  381
it          1526  390
marketing   1720  413
sales       1610  398
Chi2: 1.9322774909297424
P-value: 0.7482133205494339
There is no strong relation between Department and Atrrition


### Correlation Analysis

In [40]:
corr = df.corr(numeric_only=True, method='spearman')

attrition_corr = corr['attrition_flag'].sort_values(ascending=False)
print(attrition_corr)


fig_corr = go.FigureWidget()

fig_corr.add_trace(
    go.Heatmap(
        z=corr.values,
        x=corr.columns,
        y=corr.index,
        colorscale='RdBu_r',      
        zmin=-1,
        zmax=1,
        text=corr.values,
        texttemplate='%{text:.2f}', 
        textfont=dict(size=9, color='black'),
        hovertemplate='x: %{x}<br>y: %{y}<br>Correlation: %{z:.2f}<extra></extra>',
        colorbar=dict(title='Correlation')
    )
)

fig_corr.add_trace(
    go.Scatter(
        x=[],
        y=[],
        mode='markers',
        marker=dict(
            symbol='square',
            size=28,
            color='rgba(255,255,255,0)',
            line=dict(color='rgba(0,255,0,1)', width=2)  
        ),
        hoverinfo='skip',
        showlegend=False
    )
)

fig_corr.update_layout(
    template='plotly_white',
    title='Correlation Matrix',
    xaxis_title='',
    yaxis_title='',
    height=800,
    width=900,
    hovermode='closest',
    yaxis=dict(autorange='reversed')  
)

def highlight_hover(trace, points, selector):
    if len(points.xs) == 0:
        return
    x_val = points.xs[0]
    y_val = points.ys[0]
    fig_corr.data[1].x = [x_val]
    fig_corr.data[1].y = [y_val]

fig_corr.data[0].on_hover(highlight_hover)
fig_corr.show()



attrition_flag                   1.000000
work_life_balance                0.015167
job_involvement                  0.014934
performance_rating               0.011816
job_level                        0.011435
distance_from_home               0.011233
age                              0.007993
project_count                    0.006714
years_in_current_role            0.005144
years_at_company                 0.002583
job_satisfaction                 0.001959
absenteeism                      0.000407
monthly_income                   0.000202
training_hours_last_year        -0.000575
years_since_last_promotion      -0.001802
average_hours_worked_per_week   -0.002361
relationship_with_manager       -0.004094
hourly_rate                     -0.005518
number_of_companies_worked      -0.008217
employee_id                     -0.008766
work_environment_satisfaction   -0.011072
Name: attrition_flag, dtype: float64


FigureWidget({
    'data': [{'colorbar': {'title': {'text': 'Correlation'}},
              'colorscale': [[0.0, 'rgb(5,48,97)'], [0.1, 'rgb(33,102,172)'],
                             [0.2, 'rgb(67,147,195)'], [0.3, 'rgb(146,197,222)'],
                             [0.4, 'rgb(209,229,240)'], [0.5, 'rgb(247,247,247)'],
                             [0.6, 'rgb(253,219,199)'], [0.7, 'rgb(244,165,130)'],
                             [0.8, 'rgb(214,96,77)'], [0.9, 'rgb(178,24,43)'],
                             [1.0, 'rgb(103,0,31)']],
              'hovertemplate': 'x: %{x}<br>y: %{y}<br>Correlation: %{z:.2f}<extra></extra>',
              'text': {'bdata': ('AAAAAAAA8D8rYsIRZeyRv3/NTjtjpn' ... 'RoyoEBhz+rl5m87tOAvwAAAAAAAPA/'),
                       'dtype': 'f8',
                       'shape': '21, 21'},
              'textfont': {'color': 'black', 'size': 9},
              'texttemplate': '%{text:.2f}',
              'type': 'heatmap',
              'uid': '0cede94d-5dab-4053-b8ca-b550

###  Final decisions

#### Answer

**5 key findings:**

1. The total Attrition Rate was 19.97% (1,997 out of 10,000 employees left the company).
2. Employees who work overtime have a slightly higher attrition rate (20.07% vs 19.87%), but the difference is small and should not be treated as a strong explanation on its own.
3. Job Satisfaction and Work Life Balance scores were actually slightly `HIGHER` among employees who left, and only Work Environment Satisfaction was lower so this shows no clear, consistent trend was found across these three metrics.
4. There are small differences in attrition rate by Department (Finance 20.85%) and Job Role (Assistant 21.43%), but the Chi-Square test showed this is not statistically significant.
5. Chi-Square test result: p=0.748 (p ≥ 0.05) — no statistically significant relationship was found between Department and Attrition.

**3 business recommendations:**

1. We need to drop overtime load. To do this we must do workload disturbition across all teams.
2. Conduct quarterly employee satisfaction surveys, identify the main concerns within low-satisfaction teams, and implement targeted improvement actions based on the findings.
3. Group the main turnover drivers into key categories such as compensation, workload, career development, job satisfaction, management, and work environment, then develop targeted retention actions for the highest-impact factors.

# Additional analysis

## Workforce Structure Analysis

In [41]:
fig_workforce = make_subplots(rows= 1 ,cols=2, subplot_titles=('Count of employees by department','Count of employees per job role'))

In [42]:
dept_count = df['department'].value_counts()
role_count = df['job_role'].value_counts()

In [43]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

fig_workforce = make_subplots(rows=1, cols=2,
    subplot_titles=("Count of employees by department", "Count of employees per job role"))

fig_workforce.add_trace(go.Bar(x=dept_count.index, y=dept_count.values,
                                marker_color='#EF553B', name="department"), row=1, col=1)
fig_workforce.add_trace(go.Bar(x=role_count.index, y=role_count.values,
                                marker_color='#636EFA', name="job role"), row=1, col=2)

fig_workforce.update_layout(title_text="Workforce Structure: Department and Job Distribution",
                             showlegend=False, height=450)
fig_workforce.update_xaxes(title_text="Department", row=1, col=1)
fig_workforce.update_yaxes(title_text="Employee count", row=1, col=1)
fig_workforce.update_xaxes(title_text="Role", row=1, col=2)
fig_workforce.update_yaxes(title_text="Employee count", row=1, col=2)
fig_workforce.show()

print("Disturbution of department (%):")
print((dept_count / len(df) * 100).round(1))
print("Disturbition of role (%):")
print((role_count / len(df) * 100).round(1))

Disturbution of department (%):
department
marketing    21.3
sales        20.1
finance      19.9
hr           19.5
it           19.2
Name: count, dtype: float64
Disturbition of role (%):
job_role
analyst      25.7
assistant    25.4
executive    24.8
manager      24.1
Name: count, dtype: float64


In [44]:
job_level = pd.crosstab(df['department'],df['job_level'])

fig_job_level = px.imshow(job_level,color_continuous_scale='blues',text_auto= True,
                          labels= dict(x='job level', y= 'department',color='employee count'),
                          title = 'Department and Job level employee matrix',
                          height= 450)

fig_job_level.show()



## Salary Analysis

In [45]:
fig_salary = px.box(df, x='department', y='monthly_income', color='department',
                    color_discrete_sequence=['#08306b', "#428e08", "#96246e", "#e07033", "#93128d"],
                    title='Monthly Salary Distribution')
fig_salary.update_layout(xaxis_title="Department", yaxis_title="Monthly income", showlegend=True, height=450)

fig_salary.show()


salary_by_dep = df.groupby('department')['monthly_income'].agg(['mean','median','std']).round(2)
print(salary_by_dep)

                mean   median      std
department                            
finance     11563.93  11488.0  4871.56
hr          11574.49  11477.0  4915.58
it          11407.76  11483.5  4940.41
marketing   11295.09  11103.0  4920.63
sales       11354.72  11354.5  4982.50


In [46]:
fig_salary_level1 = px.box(df, x='job_level',
                          y = 'monthly_income',
                          color = 'job_level',
                          color_discrete_sequence=['#08306b', "#428e08", "#96246e", "#e07033", "#93128d"],
                          title='Monthly salary disturbition by Job level')

fig_salary_level1.update_layout(xaxis_title="Job level", yaxis_title="Monthly income", showlegend=True, height=450)

fig_salary_level1.show()

salary_by_level1 = df.groupby("job_level")["monthly_income"].agg(["mean", "median", "std", "count"]).round(1)
print(salary_by_level1)

              mean   median     std  count
job_level                                 
1          11384.1  11422.0  5018.8   1982
2          11482.3  11451.5  4999.7   2072
3          11608.1  11726.0  4938.4   1979
4          11311.4  10985.5  4824.8   1990
5          11396.3  11400.0  4844.4   1977


In [47]:
fig_salary_level2 = px.box(df, x='job_level', y='monthly_income',
                          color='job_role',
                          facet_col='job_role',
                          color_discrete_sequence=['#08306b', "#428e08", "#96246e", "#e07033"],
                          title='Monthly Salary Distribution by Job Level and Job Role')

fig_salary_level2.update_layout(xaxis_title="Job Level", yaxis_title="Monthly Income", showlegend=False, height=450)
fig_salary_level2.show()


salary_by_level2 = df.groupby(["job_level","job_role"])["monthly_income"].agg(["mean", "median", "std", "count"]).round(1)
print(salary_by_level2)

                        mean   median     std  count
job_level job_role                                  
1         analyst    11523.9  11644.0  5043.9    504
          assistant  11415.6  11360.5  4968.1    500
          executive  11370.7  11023.0  5006.4    472
          manager    11226.0  11407.0  5065.0    506
2         analyst    11505.6  11623.5  5034.2    532
          assistant  11368.2  11192.0  5010.2    540
          executive  11265.9  11014.0  4942.5    503
          manager    11800.3  12173.0  5007.4    497
3         analyst    11635.1  12040.0  4875.4    493
          assistant  11621.5  11898.5  5021.4    522
          executive  11577.9  11133.0  5024.2    502
          manager    11597.1  11917.0  4831.4    462
4         analyst    11428.0  11135.0  4769.0    513
          assistant  11047.2  10810.5  4715.3    482
          executive  11387.0  10985.0  4918.4    519
          manager    11371.0  10913.5  4896.1    476
5         analyst    11300.2  11054.5  4849.9 